test/unit/org/apache/cassandra/service/reads/range/ReplicaPlanMergerTest.java

```diff
@@ -0,0 +1,461 @@
+            testRanges(cl,
+                       range(max(50), max(10)),
+                       range(max(50), max(90)),
+                       range(max(90), min()),
+                       range(min(), max(10))); // wrapping range
+            testRanges(cl,
+                       range(max(25), max(15)),
+                       range(max(25), max(60)),
+                       range(max(60), min()),
+                       range(min(), max(15))); // wrapping range
+        }
+
+        // with other CLs the ranges are merged in a similar per-node basis
+        for (ConsistencyLevel cl : Arrays.asList(ALL, QUORUM, LOCAL_QUORUM, EACH_QUORUM, TWO, SERIAL, LOCAL_SERIAL))
+        {
+            testRanges(cl,
+                       range(min(), min()),
+                       range(min(), max(30)),
+                       range(max(30), max(60)),
+                       range(max(60), max(90)),
+                       range(max(90), min()));
+            testRanges(cl,
+                       range(min(), max(25)));
+            testRanges(cl,
+                       range(min(), max(40)),
+                       range(min(), max(30)),
+                       range(max(30), max(40)));
+            testRanges(cl,
+                       range(min(), max(50)),
+                       range(min(), max(30)),
+                       range(max(30), max(50)));
+            testRanges(cl,
+                       range(max(20), max(30)));
+            testRanges(cl,
+                       range(max(25), min()),
+                       range(max(25), max(30)),
+                       range(max(30), max(60)),
+                       range(max(60), max(90)),
+                       range(max(90), min()));
+            testRanges(cl,
+                       range(max(25), max(35)),
+                       range(max(25), max(30)),
+                       range(max(30), max(35)));
+            testRanges(cl,
+                       range(max(50), min()),
+                       range(max(50), max(60)),
+                       range(max(60), max(90)),
+                       range(max(90), min()));
+            testRanges(cl,
+                       range(max(50), max(10)),
+                       range(max(50), max(60)),
+                       range(max(60), max(90)),
+                       range(max(90), min()),
+                       range(min(), max(10))); // wrapping range
+            testRanges(cl,
+                       range(max(25), max(15)),
+                       range(max(25), max(30)),
+                       range(max(30), max(60)),
+                       range(max(60), max(90)),
+                       range(max(90), min()),
+                       range(min(), max(15))); // wrapping range
+        }
+    }
+
+    /**
+     * Tests range merging with a multinode cluster when the token ranges overlap between replicas.
+     */
+    @Test
+    public void testMultiNodeWithDiscontinuousRanges() throws UnknownHostException
```

Comment:
nit: Probably don't need to throw `UnknownHostException` (remember to remove its import)